src/error.rs

```diff
@@ -2,327 +2,239 @@
+	/// Get this error as remote error message.
+	///
+	/// See [`Self::remote_error()`] for more details on what a remote error is.
+	pub fn as_remote_error(&self) -> Option<&str> {
+		if let private::InnerError::RemoteError(msg) = &self.inner {
+			Some(msg)
 		} else {
-			false
+			None
 		}
 	}
-}
-
-/// An error occurred while sending an update or stream message.
-#[derive(Debug, Error)]
-#[error("{0}")]
-pub enum SendMessageError {
-	/// An I/O error occurred.
-	Io(#[from] std::io::Error),
 
-	/// The payload of the message is too large to send.
-	PayloadTooLarge(#[from] PayloadTooLarge),
-
-	/// The request has already been closed.
-	RequestClosed(#[from] RequestClosed),
-
-	/// Failed to encode the message.
-	EncodeBody(Box<dyn std::error::Error + Send>),
-}
-
-impl SendMessageError {
-	/// Check if the error is an I/O error indicating that the connection was aborted by the remote peer.
-	pub fn is_connection_aborted(&self) -> bool {
-		if let Self::Io(e) = &self {
-			e.kind() == std::io::ErrorKind::ConnectionAborted
+	/// Get this error as remote error message.
+	///
+	/// See [`Self::remote_error()`] for more details on what a remote error is.
+	pub fn into_remote_error(self) -> Option<String> {
+		if let private::InnerError::RemoteError(msg) = self.inner {
+			Some(msg)
 		} else {
-			false
+			None
 		}
 	}
 }
 
-/// An error occurred while performing a service call.
-#[derive(Debug, Error)]
-#[error("{0}")]
-pub enum ServiceCallError {
-	/// Failed to encode or send the request.
-	SendRequest(#[from] SendRequestError),
-
-	/// Failed to receive or decode the response.
-	RecvResponse(#[from] RecvMessageError),
+impl From<std::io::Error> for Error {
+	fn from(other: std::io::Error) -> Self {
+		Self::io_error(other)
+	}
 }
 
-/// An error decoding a message.
-#[derive(Debug, Error)]
-#[error("{0}")]
-pub enum FromMessageError {
-	/// The service ID was not recognized.
-	UnexpectedServiceId(#[from] crate::error::UnexpectedServiceId),
+pub(crate) mod private {
+	use super::*;
 
-	/// Failed to decode the message body.
-	DecodeBody(Box<dyn std::error::Error + Send>),
-}
+	pub(crate) fn connection_aborted() -> Error {
+		InnerError::from(std::io::Error::from(std::io::ErrorKind::ConnectionAborted)).into()
+	}
 
-// Allow a ReadMessageError to be converted to a RecvMessageError automatically.
-impl From<ReadMessageError> for RecvMessageError {
-	fn from(other: ReadMessageError) -> Self {
-		match other {
-			ReadMessageError::Io(e) => e.into(),
-			ReadMessageError::MessageTooShort(e) => e.into(),
-			ReadMessageError::InvalidMessageType(e) => e.into(),
-			ReadMessageError::PayloadTooLarge(e) => e.into(),
-		}
+	#[derive(Debug, Error)]
+	#[error("{0}")]
+	#[doc(hidden)]
+	pub enum InnerError {
+		/// An I/O error occurred.
+		Io(#[from] std::io::Error),
+
+		/// The received message is too short to be valid.
+		#[error("the message is too short to be valid: need atleast 12 bytes for the header, got only {message_len} bytes")]
```

Comment:
Worth it to remove the magic number 12 ?

Reply:
You mean not include it in the error? I prefer to have it there as proof that the lengths is indeed too low.

Reply:
I mean include it, but as a constant value defined elsewhere.

Reply:
That would mean making a custom Display implementation, which doesn't seem worth it to me.

Reply:
Although the errors are now simplified enough that maybe we could remove the `thiserror` dependency entirely :o

Reply:
Opened https://github.com/fizyr/fizyr-rpc-rs/issues/23 for it.